### PR TITLE
diffusion maps cookbook and converter factory

### PR DESCRIPTION
**Pull request by shubham808**

### CODE REVIEW

File: doc/cookbook/source/examples/converter/diffusion_maps.rst [hunk 1]
```diff
@@ -0,0 +1,37 @@
+==============================
+Diffusion Maps
+==============================
+
+Diffusion Maps is a non-linear dimensionality reduction method that uses
```
Comment (karlnapf) [inline doc/cookbook/source/examples/converter/diffusion_maps.rst]: Maybe shorten the sentence slightly.

File: doc/cookbook/source/examples/converter/diffusion_maps.rst [hunk 2]
```diff
@@ -0,0 +1,37 @@
+==============================
```
Comment (karlnapf) [inline doc/cookbook/source/examples/converter/diffusion_maps.rst]: pls adjust to text

**Comment by karlnapf** [comment]
Could you add a data file for the new example?
Otherwise good to go!

File: doc/cookbook/source/examples/converter/diffusion_maps.rst [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+Diffusion Maps
+==============================
+
+Diffusion Maps is a non-linear dimensionality reduction method that uses 
+eigenfunctions of Markov matrices to construct coordinates called diffusion maps 
+that can be used to generate efficient representations of complex geometric structures.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/converter/diffusion_maps.rst]: Is there a key equation that we could show here? You could shop around the web to see what other libraries use to describe this

**Comment by shubham808** [comment]
@karlnapf Travis is failing.... any ideas ? :)

File: doc/cookbook/source/examples/converter/diffusion_maps.rst [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+eigenfunctions of Markov matrices to diffusion maps for efficient 
+representations of complex geometric structures.
+
+For more information see :cite:`Coifman-Lafon2006Diffusionmaps`.
+
+-------
+Example
+-------
+
+we create CDenseFeatures (RealFeatures, here 64 bit float values).
```
Comment (karlnapf) [inline doc/cookbook/source/examples/converter/diffusion_maps.rst]: Capital W

File: doc/cookbook/source/examples/converter/diffusion_maps.rst [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+.. sgexample:: diffusionmaps.sg:create_features
+
+We create the :sgclass:`CDiffusionMaps` instance, and set its parameters. The diffusion kernel :math: `k` must satisfy
```
Comment (karlnapf) [inline doc/cookbook/source/examples/converter/diffusion_maps.rst]: We create "a".
Also pls start new sentences in new lines

File: doc/cookbook/source/examples/converter/diffusion_maps.rst [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+the following properties:
+
+1. :math: `k` is symmetric :math: `{\bf k}(x, y) = {\bf k}(y, x)`
```
Comment (karlnapf) [inline doc/cookbook/source/examples/converter/diffusion_maps.rst]: this should be said in the intro

File: doc/cookbook/source/examples/converter/diffusion_maps.rst [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+2. :math: `k` is positivity preserving :math: `{\bf k}(x, y) ≥ 0`
+
+.. sgexample:: diffusionmaps.sg:set_parameters
+
+Then we apply DiffusionMaps, which gives us the distance embeddings.
```
Comment (karlnapf) [inline doc/cookbook/source/examples/converter/diffusion_maps.rst]: remove "the"

Also, please spell it "diffusion maps"

**Comment by karlnapf** [comment]
LGTM.

@vinx13 did you change any of the API here? Do we need to change anything for that?

**Comment by karlnapf** [comment]
No idea why windows fails, restarted.
Travis is OK so all code seems good

**Comment by vinx13** [comment]
Yes, we have new api in feature/transformers branch. Something like 
`features = converter.transform(features)`
And there is factory method `transformer`.

**Comment by karlnapf** [comment]
Ah ok that is a feature branch then.
So we will merge this, and then refactor upon feature branch merging

**Comment by karlnapf** [comment]
Very useful stuff, thanks! :)

**Pull request merged**